wgpu-hal/src/vulkan/adapter.rs

```diff
@@ -132,32 +158,30 @@ impl PhysicalDeviceFeatures {
                         )
                         .descriptor_indexing(requested_features.intersects(indexing_features()))
                         .shader_sampled_image_array_non_uniform_indexing(
-                            requested_features.contains(
-                                wgt::Features::TEXTURE_BINDING_ARRAY
-                                    | wgt::Features::SAMPLED_TEXTURE_AND_STORAGE_BUFFER_ARRAY_NON_UNIFORM_INDEXING,
-                            ),
+                            needs_sampled_image_non_uniform,
                         )
                         .shader_storage_image_array_non_uniform_indexing(
-                            requested_features.contains(
-                                wgt::Features::TEXTURE_BINDING_ARRAY
-                                    | wgt::Features::STORAGE_RESOURCE_BINDING_ARRAY
-                                    | wgt::Features::UNIFORM_BUFFER_AND_STORAGE_TEXTURE_ARRAY_NON_UNIFORM_INDEXING,
-                            ),
+                            needs_storage_image_non_uniform,
                         )
-                        //.shader_storage_buffer_array_non_uniform_indexing(
                         .shader_uniform_buffer_array_non_uniform_indexing(
-                            requested_features.contains(
-                                wgt::Features::BUFFER_BINDING_ARRAY
-                                    | wgt::Features::UNIFORM_BUFFER_AND_STORAGE_TEXTURE_ARRAY_NON_UNIFORM_INDEXING,
-                            ),
+                            needs_uniform_buffer_non_uniform,
                         )
                         .shader_storage_buffer_array_non_uniform_indexing(
-                            requested_features.contains(
-                                wgt::Features::BUFFER_BINDING_ARRAY
-                                    | wgt::Features::STORAGE_RESOURCE_BINDING_ARRAY
-                                    | wgt::Features::SAMPLED_TEXTURE_AND_STORAGE_BUFFER_ARRAY_NON_UNIFORM_INDEXING,
-                            ),
+                            needs_storage_buffer_non_uniform,
+                        )
+                        .descriptor_binding_sampled_image_update_after_bind(
+                            private_caps.update_after_bind,
```

Comment:
An ideal implementation would have something like a mask:
```rust
bitflags {
bitflags::bitflags!(
    pub struct UpdateAfterBindDescriptorTypes: u32 {
        const SAMPLED_IMAGE = 0x1;
        const STORAGE_IMAGE = 0x2;
        const UNIFORM_BUFFER = 0x4;
        const STORAGE_BUFFER = 0x8;
    }
);
```
We can afford to be more detailed here because it's not user-visible directly. As we discussed, this can be figure out by the backend based on the requested limits.

So if we aren't implementing it this way, let's write down a note somewhere with a TODO.